File: builddeps/builddeps.kt
import java.util.*

fun main() {
    val lines = generateSequence { readLine() }.toList()

    // inputs
    val N = lines[0].toInt()

    val graph = mutableMapOf<String, List<String>>()
    for (i in 1 until N + 1) {
        val line = lines[i].trim()
        val (key, neighbours) = line.split(":")
        if (neighbours.isNotEmpty()) {
            graph[key] = neighbours.trim().split(" ")
        }
    }

    val start = lines[N + 1].trim()

    fun invertMap(map: Map<String, List<String>>): Map<String, List<String>> {
        val invertedMap = mutableMapOf<String, MutableList<String>>()
        for ((key, neighbours) in map) {
            for (neighbour in neighbours) {
                if (!invertedMap.containsKey(neighbour)) {
                    invertedMap[neighbour] = mutableListOf(key)
                } else {
                    invertedMap[neighbour]?.add(key)
                }
            }
        }
        return invertedMap
    }


    // Step 1 is to invert the map
    val invertedGraph = invertMap(graph)

    // Step 2 is to do a topological sort (reference: Neetcode 210)
    fun toposort(graph: Map<String, List<String>>, start: String): List<String> {
        val output = mutableListOf<String>()
        val visited = mutableSetOf<String>()

        tailrec fun dfs(node: String) {
            if (node in visited) {
                return
            }
            for (neighbour in graph[node] ?: emptyList()) {
                dfs(neighbour)
            }

            visited.add(node)
            output.add(node)
        }

        dfs(start)
        return output.reversed()
    }

    for (i in toposort(invertedGraph, start)) {
        println(i)
    }
}

File: builddeps/builddeps.py
import sys
sys.setrecursionlimit(10**6)
lines = sys.stdin.read().splitlines()

# inputs
N = int(lines[0])

graph = {}
for i in range(N):
    line = lines[1 + i].strip()
    key, neighours = line.split(":")
    if neighours != "":
        graph[key] = neighours.strip().split(" ")


start = lines[N + 1]

# remove nodes without any dependencies


def invert_map(map):
    inverted_map = {}
    for key, neighbours in map.items():
        for neighbour in neighbours:
            if neighbour not in inverted_map:
                inverted_map[neighbour] = [key]
            else:
                inverted_map[neighbour].append(key)
    return inverted_map


inverted_graph = invert_map(graph)

# topological sort
def toposort(graph):
    output = []
    visited = set()

    def dfs(node):
        if node in visited:
            return
        for neighour in graph.get(node, []):
            dfs(neighour)

        visited.add(node)
        output.append(node)

    dfs(start)
    return output


for i in reversed(toposort(inverted_graph)):
    print(i)

# code works but only if you increase stack limit.
